### PR TITLE
Updates to proto pathway for Schemas

### PR DESCRIPTION
**Pull request by alxrsngrtn**
- Added a new FieldType to Kolin Schemas: `Tuple`.
- Updated Schema normalization (part of hash) to support tuples. 
- Undid adding refinements to the schema proto -- tests revealed that it added redundant information. 
- Decoded collections, references, and tuples on Schema fields into Kotlin

### CODE REVIEW

File: src/runtime/schema.ts [hunk 1]
```diff
@@ -321,6 +321,8 @@ export class Schema {
         str += '[&(' + schema.schema.model.entitySchema.normalizeForHash() + ')]';
       } else if (kind === 'schema-collection' && schema.kind === 'schema-primitive') {
         str += '[' + schema.type + ']';
+      } else if (kind === 'schema-tuple') {
+        str += `(${types.map(t => t.type).join('|')})`
```
Comment (alxrsngrtn) [inline src/runtime/schema.ts:325]: @mykmartin How does this look to you?

Comment (alxrsngrtn) [reply src/runtime/schema.ts:325]: (Test case below.)

File: java/arcs/core/data/proto/TypeProtoDecoders.kt [hunk 2]
```diff
@@ -30,7 +30,15 @@ fun PrimitiveTypeProto.decode() = when (this) {
 }
 
 /** Converts a [PrimitiveTypeProto] protobuf instance into a Kotlin [FieldType] instance. */
-fun PrimitiveTypeProto.decodeAsFieldType(): FieldType.Primitive = FieldType.Primitive(decode())
+fun PrimitiveTypeProto.decodeAsFieldType() = FieldType.Primitive(decode())
+
+/** Converts a [ReferenceTypeProto] protobuf instance into a Kotlin [FieldType] instance. */
+fun ReferenceTypeProto.decodeAsFieldType() = FieldType.EntityRef(decode().entitySchema!!.hash)
```
Comment (alxrsngrtn) [inline java/arcs/core/data/proto/TypeProtoDecoders.kt:36]: I found it difficult to wrap the `.hash` accessor with `requireNotNull`. Any ideas for a better way?

File: src/tools/manifest2proto.ts [hunk 3]
```diff
@@ -242,7 +242,7 @@ export async function schemaToProtoPayload(schema: Schema) {
     names: schema.names,
     fields: objectFromEntries(await Promise.all(Object.entries(schema.fields).map(
       async ([key, value]) => [key, await schemaFieldToProtoPayload(value)]))),
-    hash: await schema.hash()
+    hash: await schema.hash(),
```
Comment (alxrsngrtn) [inline src/tools/manifest2proto.ts:245]: TODO(self) remove comma.

File: java/arcs/core/data/proto/SchemaProtoDecoder.kt [hunk 4]
```diff
@@ -24,21 +24,22 @@ fun SchemaProto.decodeNames() = getNamesList().map { SchemaName(it) }.toSet()
...
     val collections = mutableMapOf<FieldName, FieldType>()
-    for ((name, type) in getFieldsMap()) {
-        when (type.getDataCase()) {
+    for ((name, type) in fieldsMap) {
+        when (type.dataCase) {
             TypeProto.DataCase.PRIMITIVE -> singletons[name] = type.decodeAsFieldType()
-            // TODO: TypeProto.DataCase.COLLECTIONS
-            // TODO: TypeProto.DataCase.REFERENCE
+            TypeProto.DataCase.REFERENCE -> singletons[name] = type.decodeAsFieldType()
+            TypeProto.DataCase.TUPLE -> singletons[name] = type.decodeAsFieldType()
```
Comment (bgogul) [inline java/arcs/core/data/proto/SchemaProtoDecoder.kt:31]: Since everything updates `singletons` map, how about?
```suggestion
             TypeProto.DataCase.REFERENCE, 
             TypeProto.DataCase.PRIMITIVE, 
             TypeProto.DataCase.TUPLE -> singletons[name] = type.decodeAsFieldType()
```

File: java/arcs/core/data/proto/TypeProtoDecoders.kt [hunk 5]
```diff
@@ -30,7 +30,15 @@ fun PrimitiveTypeProto.decode() = when (this) {
...
+
+/** Converts a [TupleTypeProto] protobuf instance into a Kotlin [FieldType] instance. */
+fun TupleTypeProto.decodeAsFieldType(): FieldType.Tuple = FieldType.Tuple(
```
Comment (bgogul) [inline java/arcs/core/data/proto/TypeProtoDecoders.kt:39]: `...decodeAsFieldType() = FieldType.Tuple(elementsList.map { it.decodeAsFieldType() })`

File: java/arcs/core/data/proto/TypeProtoDecoders.kt (hunk 2 above)
Comment (bgogul) [reply java/arcs/core/data/proto/TypeProtoDecoders.kt:36]: Doesn't `requireNotNull(decode().entitySchema).hash` work? 

I think it might good to be a explicit here:
```
fun ReferenceTypeProto.decodeAsFieldType(): FieldType.EntityRef {
   val entitySchema = requireNotNull(decode().entitySchema) { 
        "Field that is a reference to an non-entity type is not possible." 
   } 
   return FieldType.EntityRef(entitySchema.hash)
}
```

File: javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt [hunk 6]
```diff
@@ -52,4 +52,150 @@ class SchemaProtoDecoderTest {
...
+        }
+        """.trimIndent()
+        val schema = decodeSchemaProtoText(schemaProtoText)
+        assertThat(schema.fields.collections).isEqualTo(
+            mapOf("text" to FieldType.Text, "bool" to FieldType.Boolean))
+    }
+
+    @Test
+    fun decodesCollectionsOfReferencesInSchemaProto() {
+        val schemaProtoText = """
```
Comment (bgogul) [inline javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt:87]: Can you also add the Arcs schema as a comment here and else where? It is kind of hard to parse the proto.

File: java/arcs/core/data/proto/SchemaProtoDecoder.kt (hunk 4 above)
Comment (alxrsngrtn) [reply java/arcs/core/data/proto/SchemaProtoDecoder.kt:31]: Nice, I like this!

File: java/arcs/core/data/proto/TypeProtoDecoders.kt (hunk 2 above)
Comment (alxrsngrtn) [reply java/arcs/core/data/proto/TypeProtoDecoders.kt:36]: I like that err msging.

File: java/arcs/core/data/proto/TypeProtoDecoders.kt (hunk 5 above)
Comment (alxrsngrtn) [reply java/arcs/core/data/proto/TypeProtoDecoders.kt:39]: I need to add the return type `FieldType.Tuple` in order to solve an infinite type recursion problem in Kotlin Type Inference. 

When I remove this, I get errors on line 56 (decoding a Tuple as field type).

**Comment by bgogul** [comment]
@alxrsngrtn, some comments are resolved, but I don't see any related changes in the PR? Just checking to make sure that I did not miss anything.

File: javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt (hunk 6 above)
Comment (alxrsngrtn) [reply javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt:87]: 👍

**Comment by bgogul** [comment]
> @alxrsngrtn, some comments are resolved, but I don't see any related changes in the PR? Just checking to make sure that I did not miss anything.

Never mind. I spoke too soon. :)

File: javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt [hunk 7]
```diff
@@ -22,6 +22,10 @@ fun decodeSchemaProtoText(protoText: String): Schema {
 
 @RunWith(JUnit4::class)
 class SchemaProtoDecoderTest {
+
```
Comment (bgogul) [inline javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt:25]: Can you move these comments to inside the tests where you declare the text protos?

Comment (alxrsngrtn) [reply javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt:25]: Inside the test methods?

**Comment by alxrsngrtn** [comment]
> > @alxrsngrtn, some comments are resolved, but I don't see any related changes in the PR? Just checking to make sure that I did not miss anything.
> 
> Never mind. I spoke too soon. :)

I often mark stuff resolve while implementing the feedback in the PR so I can keep track of what I have to do. This does mean that stuff is resolved before it's pushed :)

File: javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt (hunk 7 above)
Comment (bgogul) [reply javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt:25]: I meant something like this, instead of annotating the test function:
```suggestion
    @Test
    fun decodesNamesInSchemaProto() {
         // schema Thing, Object
         val schemaProtoText = """
           names:  "Thing"
           names:  "Object"
         """.trimIndent()
         ...
    }
```

File: javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt [hunk 8]
```diff
@@ -52,4 +57,160 @@ class SchemaProtoDecoderTest {
...
+        fields {
+          key: "bool"
+          value: {
+            collection {
+              collection_type: {
+                primitive: BOOLEAN
+              } 
+            }
+          }
+        }
```
Comment (mykmartin) [inline javatests/arcs/core/data/proto/SchemaProtoDecoderTest.kt:86]: Consider writing in a more compact form? I think this could be more readable, especially for the larger examples further down:

```
fields {
  key: "text"
  value: {
    collection { collection_type: { primitive: TEXT } }
  }
}
fields {
  key: "bool"
  value: {
    collection { collection_type: { primitive: BOOLEAN } }
  }
}
```
